blueprint/core/blueprint_View.cpp

```diff
@@ -144,37 +185,70 @@ namespace blueprint
         auto w = cachedFloatBounds.getWidth();
         auto h = cachedFloatBounds.getHeight();
 
-        dispatchViewEvent("onMeasure", w, h);
+        dispatchViewEvent("onMeasure", detail::makeViewEventObject({
+            {"width", w},
+            {"height", h}
+        }));
     }
 
     void View::mouseDown (const juce::MouseEvent& e)
     {
-        dispatchViewEvent("onMouseDown", e.x, e.y);
+        dispatchViewEvent("onMouseDown", detail::makeViewEventObject(e));
     }
 
     void View::mouseUp (const juce::MouseEvent& e)
     {
-        dispatchViewEvent("onMouseUp", e.x, e.y);
+        dispatchViewEvent("onMouseUp", detail::makeViewEventObject(e));
     }
 
     void View::mouseDrag (const juce::MouseEvent& e)
     {
-        auto mouseDown = e.mouseDownPosition;
-        dispatchViewEvent("onMouseDrag", e.x, e.y, mouseDown.x, mouseDown.y);
+        // TODO: mouseDrag isn't a dom event... is it?
+        dispatchViewEvent("onMouseDrag", detail::makeViewEventObject(e));
     }
 
     void View::mouseDoubleClick (const juce::MouseEvent& e)
     {
-        dispatchViewEvent("onMouseDoubleClick", e.x, e.y);
+        dispatchViewEvent("onMouseDoubleClick", detail::makeViewEventObject(e));
     }
 
     bool View::keyPressed (const juce::KeyPress& key)
     {
-        dispatchViewEvent("onKeyPress", key.getKeyCode());
+        dispatchViewEvent("onKeyPress", detail::makeViewEventObject(key));
+
+        // We always inform the underlying juce::Component that we've consumed the event,
+        // because we manually bubble a SyntheticEvent wrapper up the javascript view tree.
+        // However, because the React app may be only a subtree of the overall app architecture,
+        // we skip up to the ReactApplicationRoot parent and restart the keypress event propagation
+        // up there.
+        if (auto* parent = findParentComponentOfClass<ReactApplicationRoot>())
```

Comment:
Struggling to wrap my head around this part. Could you give a little more explanation?

Reply:
Yep, so basically juce does keypress event bubbling itself, walking up the `Component*` heirarchy. But we do that manually ourselves inside the juce-blueprint backend so that we can better shim the UIEvent spec, so we return `true` here to make sure juce's Component implementation stops, otherwise we'd get event bubbling twice. Now, if we just simply stopped here, any `juce::Component` starting with ReactApplicationRoot or higher will never receive keypress events that originated somewhere in the React tree, but our js bubbling will stop at the first child of the react root (because it knows of no parent higher). That means, for example, we might catch a Ctrl+D keypress on some component in the react tree, bubble it up the javascript heirarchy where maybe nothing happens, and then stop. That would be undesired, because we want ReactAppRoot to still be able to catch that and enable debugging. So this line, 224, is at a point where we've already completed our view event bubbling in js land, we don't want juce::Component to bubble the same event to the same tree twice, but we do want it to bubble starting from the ReactAppRoot.

Reply:
Awesome. Makes total sense. Thanks.